source/fundamentals/crud/compound-operations.txt

```diff
@@ -0,0 +1,313 @@
+MongoDB Java driver. 
+
+Compound operations consist of a read and write operation performed as one
+**atomic operation**. An atomic operation is an operation which either completes
+entirely, or does not complete at all. Atomic operations cannot partially complete.
+
+Atomic operations can help you avoid **race conditions** in your code. A
+race condition occurs when your code's behavior is dependent on the order of
+uncontrollable events.
+
+MongoDB supports the following compound operations:
+
+- Find and update one document
+- Find and replace one document
+- Find and delete one document
+
+If you need to perform more complex tasks atomically, such as reading and
+writing to more than one document, use **transactions**. Transactions are a
+feature of MongoDB and other databases that lets you define an arbitrary
+sequence of database commands as an atomic operation.  
+
+For more information on atomic operations and atomicity, see 
+:manual:`the MongoDB manual entry for atomicity and transactions </core/write-operations-atomicity/>`.
+
+For more information on transactions, 
+:manual:`see the MongoDB manual entry for transactions </core/transactions/>`.
+
+How to Use Compound Operations
+------------------------------
+
+This section shows how to use each compound operation with the MongoDB Java Driver.
+
+The following examples use a collection containing these two sample documents.
+
+.. code-block:: json
+
+    {"_id": 1, "food": "donut", "color": "green"}
+    {"_id": 2, "food": "pear", "color": "yellow"}
+
+`The full code for the following examples is available on Github here. <https://github.com/mongodb/docs-java/blob/master/source/includes/fundamentals/code-snippets/CompoundOperatorsIndividualExamples.java>`__
+
+.. note:: Before or After the Write?
+
+   By default, compound operations return your found document in the state
+   before your write operation. You can retrieve your found document in the
+   state after your write operation by using the options class corresponding to
+   your compound operation. You can see an example of this configuration in the 
+   :ref:`Find and Replace example below <java-find-and-replace-example>`.
+
+Find and Update
+~~~~~~~~~~~~~~~
+
+To find and update one document, use the ``findOneAndUpdate()`` method of the
+``MongoCollection`` class. The ``findOneAndUpdate()`` method returns your found
+document or ``null`` if no documents match your query.
+
+Example
+^^^^^^^
+
+The following example uses the ``findOneAndUpdate()`` method to find a
+document with the ``color`` field set to ``"green"`` and  update the
+``food`` field in that document to ``"pizza"``. 
+
+The example also uses a ``FindOneAndUpdateOptions`` instance to specify the
+following options:
+
+- Exclude the ``_id`` field from the found document.
+- Specify to insert a document if no documents match the query. MongoDB and
+  other databases refer to this operation as an upsert.
```

Comment:
```suggestion
- Specify an upsert, which inserts a document if no documents match the query.
```
Suggestion: It's more clear if you define the term after it is mentioned. As is, the user doesn't know what it's called till they get to the second sentense.

Reply:
This comment was discussed synchronously. At first I did not agree with this edit, but I after discussion I agreed that this edit, paired with a link to a full description of upsert operations, improves the page.  

Fix: